### ISSUE TITLE
Remove RECOMMENDED ENGINE from documentation or provide arguments

### ISSUE THREAD

**Issue by gentunian** [FriendsOfCake/CakePdf#189]
Hey there :)

**TL;DR:** _Please, don't recommend this engine. Is a PITA put this working and may depend on the environment if this works or not. It may scare away people of this plugin if they don't research like me why `wkhtmltopdf` is not working._

I've started using this plugin but I found a lot of problems regarding `wkhtmltopdf`. I spent most of my time trying to making it work just because is recommended. But honestly, It would be nice to read why is recommended. I needed to compile the source manually and creating a custom `.deb` package because none of the `wkhtmltopdf` precompiled versions worked:
- wkhtmltopdf 0.12.3 (Fedora 23, 32 bits binary version)
- wkhtmltopdf 0.12.3 (Fedora 23, 64 bits binary version)
- wkhtmltopdf 0.12.4 ( Ubuntu 16.04 64 bits version)

I've researched the issue debugging the code a lot, trying to find out if `CakePdf` code was wrong, If my code was wrong, or if the `proc_open()` method was being used incorrectly. I couldn't find anything.

The only reason why I've decided to build and compile the source by my own was I got tired of reading this message from `wkhtmltopdf` output:

> Reduced Functionality:
>   This version of wkhtmltopdf has been compiled against a version of QT without
>   the wkhtmltopdf patches. Therefore some features are missing, if you need
>   these features please use the static version.
> 
>   Currently the list of features only supported with patch QT includes:
> - Printing more than one HTML document into a PDF file.
> - Running without an X11 server.
> - Adding a document outline to the PDF file.
> - Adding headers and footers to the PDF file.
> - Generating a table of contents.
> - Adding links in the generated PDF file.
> - Printing using the screen media-type.
> - Disabling the smart shrink feature of webkit.

Despite that I have tested correctly the binary with custom commands and in my daily box, just like:

``` bash
wkhtmltopdf --cookie 'csrftoken' '5158f8282ce2c072aaf17f34a4a84975afec4091a68b585e18a0ee042b5ced5b7baaee53f06eadb0cadcb33e1c0051c801c79baccd5195628a21e079418ab868' --cookie 'remember_me' 'Q2FrZQ%3D%3D.N2VjMjY1YTc3OTUyYzUwOTUyZTQ2MGYzNGEzNGE3NzJhYjUxZDEwZDJhMjVjMGZlMDY5ZTAxNjIwMGQyYmEzOIsXyLTI2kpnma%2Fep7lh52jYGUDN0DRQCw9zNcAb8gJ47Set%2FLhZt3S1%2FCuuUASvsq1FAvPObKk4pSWTRClH11f%2FrIZzwslbd%2FOwt6ergrez2mIE5IVNzr9CoCLI0WpS2Zd2oPFqtGMIhJJR4fjRM1nmZ51knudEifhdfTXyRUwJZzRxoBuM6GbQKBot7fqtW88xnEEDIR4m%2BNONCdSZFVEyZyAhKo3F5Grj5i44pTfK9z3pmHlqJzPZsZt3hHVinw%3D%3D' --cookie 'CAKEPHP' 'd26kmgrtasi9og6ba3g4j6pvh0' --post username 'developer' --post password '123' "http://my.example.com/report.pdf" blah.pdf
```

and:

``` bash
/usr/bin/wkhtmltopdf --cookie-jar cookies.txt "http://my.example.com/report.pdf" blah.pdf
```

and ([this is actually how CakePdf invokes wkhtmltopdf, that is, piping the HTML to the process](https://github.com/FriendsOfCake/CakePdf/blob/master/src/Pdf/Engine/WkHtmlToPdfEngine.php#L64)):

``` bash
echo "<h1>title</h1>" | /usr/bin/wkhtmltopdf --quiet --print-media-type --orientation 'portrait' --page-size 'A4' --encoding 'UTF-8' - - > blah.pdf
```

All of them worked correctly, in my box and in the production server. Nevertheless, it won't work with `CakePdf` plugin (binary path was right). After compiling and building my own `wkhtmltopdf` binary - that took 1 hour and a half -, I managed to get it working but at an extreme CPU cost. CPU went to 100% and for just printing _TEST_ it will take almost 2 minutes.

So I've decided to use `DomPdf` that worked just fine but with issues with javascript. After that, I tried to do something about cpu usage but I didn't find anything. So now I've decided to use `Tcpdf`. I didn't test it already with javascript. But please, provide a decent arguments why you recommend using `wkhtmltopdf` engine where all I've found is people having trouble with it.

Take this comment as a brief story of someone being stuck by a day just for trying the recommended engine :)

**Comment by Spriz** [comment]
I think one of the best arguments is the performance improvements in wkhtmltopdf - but it also involves extra complexity of adding another binary executable as a dependency.

**Comment by gentunian** [comment]
@Spriz thanks for your feedback. Performance is awful. Can you please share info or documentation about what you are saying? I really want to implement the *recommended engine* if it really worth it.

**Comment by dakota** [comment]
Wkhtmltopdf is blisteringly fast compared to any other pdf generators. Reason is quite simply because it is written in C and uses webkit to render the HTML. The other supported engines are all written in pure PHP and only support a subset of HTML and CSS.

Try version 0.12.1 (http://download.gna.org/wkhtmltopdf/0.12/0.12.1/). It seems like the wkhtmltopdf maintainers have stopped releasing standalone versions for some reason.

**Comment by gentunian** [comment]
@dakota great! downloading 0.12.2 binary worked nice. Thanks for the info. Seems like my build was not performance at all. I hope this issue will serve to someone like me having issues with `wkhtmltopdf`.

**Comment by dereuromark** [comment]
Maybe you can help adding a small note in the documentation/readme here?

**Comment by gentunian** [comment]
Of course. Would you prefer referencing the issue? or just commenting that is recommended to use the precompiled packages?

**Comment by gentunian** [comment]
I'm not sure where to start. I've cloned the repo and created a branch `update-readme`. For creating PR I need to set a branch in github but i'm not sure to `git push -u oriding update-readme`. Is that how you manage and collaborate in this project? I don't want to break any rule :)

**Comment by berarma** [comment]
On Debian and derivatives there's a xvfb wrapper that can help. Write a script at /usr/local/bin/wkhtmltopdf with this contents:

```
#!/bin/bash
exec /usr/bin/xvfb-run -- /usr/bin/wkhtmltopdf "$@"
```

Set execution permissions. Configure CakePDF to use this script. Still missing some features only present in the static version but it won't complain anymore about X server not found.

**Comment by ADmad** [comment]
The latest release will give you a more detailed error message when wkhtmltopdf fails to generate pdf. It upto you to figure out how to fix the issue.
